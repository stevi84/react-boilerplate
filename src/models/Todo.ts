import { Components } from '../apis/api';

export type Todo = Components.Schemas.Todo;

export const getEmptyTodo = (): Todo => ({
  id: 0,
  tsCreate: '',
  tsUpdate: '',
  owner: '',
  dueDate: '',
  description: '',
  completed: false,
});
