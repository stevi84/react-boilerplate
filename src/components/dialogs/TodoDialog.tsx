import { useEffect, useState } from 'react';
import { useTranslation } from 'react-i18next';
import { Locale } from '../../globals/Translations';
import { getEmptyTodo, Todo } from '../../models/Todo';
import { useAppDispatch, useAppSelector } from '../../reducers/Store';
import { deleteTodo, readTodos, todosSelector } from '../../reducers/TodosReducer';
import { ModalConfirmCancel } from '../common/ModalConfirmCancel';
import { DataTable } from '../dataTable/DataTable';
import { useNavigate } from 'react-router-dom';
import { Paper } from '@mui/material';
import { formatDate } from '../../globals/Formatters';
import { MainLayout } from '../common/MainLayout';

export const TodoDialog = () => {
  const todos: Todo[] = useAppSelector(todosSelector);
  const [deleteModalState, setDeleteModalState] = useState({ visible: false, entityId: -1 });
  const { t, i18n } = useTranslation();
  const lang: Locale = i18n.language as Locale;
  const navigate = useNavigate();
  const dispatch = useAppDispatch();

  useEffect(() => {
    if (todos.length === 0) {
      dispatch(readTodos());
    }
  }, [dispatch, todos]);

  return (
    <MainLayout>
      <Paper sx={{ margin: 1 }}>
        <DataTable<Todo>
          id={'todo-table'}
          columns={[
            {
              field: 'owner',
              headerName: t('owner'),
              cellDataType: 'text',
            },
            {
              field: 'dueDate',
              headerName: t('dueDate'),
              cellDataType: 'text',
              valueFormatter: (params) => formatDate(params.value, lang),
            },
            {
              field: 'description',
              headerName: t('description'),
              cellDataType: 'text',
            },
            { field: 'completed', headerName: t('completed'), cellDataType: 'boolean' },
          ]}
          rowsData={todos}
          manager={{
            create: () => navigate('/todos/new'),
            read: () => dispatch(readTodos()),
            update: (entity: Todo) => navigate(`/todos/${entity.id}`),
            delete: (entity: Todo) => setDeleteModalState({ visible: true, entityId: entity.id }),
            getEmpty: getEmptyTodo,
          }}
        />
        <ModalConfirmCancel
          isOpen={deleteModalState.visible}
          headerText={t('delete')}
          bodyText={t('dialog_delete_text')}
          confirmButtonText={t('delete')}
          cancelButtonText={t('cancel')}
          confirmAction={() => {
            dispatch(deleteTodo(deleteModalState.entityId));
            setDeleteModalState({ visible: false, entityId: -1 });
          }}
          cancelAction={() => setDeleteModalState({ visible: false, entityId: -1 })}
        />
      </Paper>
    </MainLayout>
  );
};
