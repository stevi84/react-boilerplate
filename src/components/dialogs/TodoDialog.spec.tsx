import { describe, it, expect, vi } from 'vitest';
import { ShallowRenderer, createRenderer } from 'react-test-renderer/shallow';
import { render } from '@testing-library/react';
import { todo1 } from '../../../test/data/Todo';
import { TodoDialog } from './TodoDialog';

vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (str: string) => str,
    i18n: {
      language: 'de',
      changeLanguage: vi.fn(),
    },
  }),
}));

vi.mock('react-router-dom', () => ({ useNavigate: () => vi.fn() }));

vi.mock('../../hooks/UseNotifier', () => ({ useNotifier: vi.fn() }));

const { useAppSelectorMock, dispatchMock } = vi.hoisted(() => ({
  useAppSelectorMock: vi.fn(),
  dispatchMock: vi.fn(),
}));
vi.mock('../../reducers/Store', () => ({
  useAppSelector: useAppSelectorMock,
  useAppDispatch: () => dispatchMock,
}));

describe('TodoDialog', () => {
  it('should equal saved snapshot', () => {
    useAppSelectorMock.mockReturnValueOnce([todo1]).mockReturnValue(false);
    const renderer: ShallowRenderer = createRenderer();
    renderer.render(<TodoDialog />);
    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('should load todos if not present', () => {
    useAppSelectorMock.mockReturnValueOnce([]).mockReturnValue(false);
    expect(dispatchMock.mock.calls.length).toEqual(0);
    render(<TodoDialog />);
    expect(dispatchMock.mock.calls.length).toEqual(1);
    //expect(dispatchMock.mock.calls[0][0]).toEqual(readTodos('de'));
  });
});
