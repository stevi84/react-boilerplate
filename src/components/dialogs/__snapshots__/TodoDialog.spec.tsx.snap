// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoDialog > should equal saved snapshot 1`] = `
<MainLayout>
  <ForwardRef(Paper)
    sx={
      {
        "margin": 1,
      }
    }
  >
    <DataTable
      columns={
        [
          {
            "cellDataType": "text",
            "field": "owner",
            "headerName": "owner",
          },
          {
            "cellDataType": "text",
            "field": "dueDate",
            "headerName": "dueDate",
            "valueFormatter": [Function],
          },
          {
            "cellDataType": "text",
            "field": "description",
            "headerName": "description",
          },
          {
            "cellDataType": "boolean",
            "field": "completed",
            "headerName": "completed",
          },
        ]
      }
      id="todo-table"
      manager={
        {
          "create": [Function],
          "delete": [Function],
          "getEmpty": [Function],
          "read": [Function],
          "update": [Function],
        }
      }
      rowsData={
        [
          {
            "completed": true,
            "description": "Weihnachtsgeschenke besorgen",
            "dueDate": "2022-12-24T08:00:00",
            "id": 1,
            "owner": "Kerstin Barth",
            "tsCreate": "2022-12-01T15:02:28",
            "tsUpdate": "2022-12-01T15:02:28",
          },
        ]
      }
    />
    <ModalConfirmCancel
      bodyText="dialog_delete_text"
      cancelAction={[Function]}
      cancelButtonText="cancel"
      confirmAction={[Function]}
      confirmButtonText="delete"
      headerText="delete"
      isOpen={false}
    />
  </ForwardRef(Paper)>
</MainLayout>
`;
